### DISCUSSION TITLE
Is there a way to have Vercel for Github wait for Github Action to complete before deploying?

### DISCUSSION THREAD

**Discussion by aegatlin** [vercel/community#390] [Show & Tell]
I want to run my GH action tests on my main branch, and if they fail, I don't want Vercel to deploy to prod. At the moment it will deploy to prod before the test suite even runs.

## Additional Context (skippable)

The only workflow that seems "simple" for Vercel is a branch-based paradigm, where my GH-action tests run against the PR and I get the PR previews. That workflow makes the changes feel safe, so that's great! But, I would also like to have a trunk-based paradigm be "simple" as well. So, I push small changes directly to main (i.e., the prod branch) and have them run through a GH-action tests before auto-deploying via Vercel-for-Github. If the tests fail, Vercel doesn't deploy. I'm imagining some setting somewhere that is like `waitForActions: true`, so I don't have to use a 3rd-party vercel deploy action that uses the deploy hooks, etc., but is instead a first-party solution. So, my questions are: (1) is there a "simple" trunk-based solution already? (2) If not, are there plans for one? (3) Am I missing some simpler solution, somehow?

I love Vercel and what yall are doing. Keep it up! :D <3

**Comment by criess** [comment]
see my comment on concurrent CI here: https://github.com/vercel/community/discussions/285#discussioncomment-2242832

On the linked thread a workaround is proposed … (Github actions issueing deploy hook and only then build not from code push/update)

**Reply by aegatlin** [reply]
Thanks for the reply! I had been reading that thread previously but didn't make it too far down because hacking an IgnoreBuildStep script to bail on a failed test suite was not something I wanted to burn hours figuring out if I could make it work. But it was where I saw the whole deploy hook strategy. So, are you under the impression that the Vercel recommended approach is to turn off auto-deploys and use their deploy hooks? It's at least a technically feasible task, I'm just hoping there's a more tailored approach that's in-line with the typical Vercel world-class developer experience I've come to appreciate.

Another failed approach I took was to hack my `npm run build` task to try and run my tests, but since Vercel deploys only the production bundle, that didn't work. But, maybe if I was using Vite or something, deployed on Vercel, I could hack that build step. Not sure... maybe a feature request is an option. Something like a "prebuild command" in the settings, and which gives you access to your whole source code, tests and all.
